NorthstarDLL/server/servernethooks.cpp

```diff
@@ -0,0 +1,208 @@
+	{
+		if (!Cvar_net_debug_atlas_packet_insecure->GetBool())
+		{
+			if (dbg)
+				spdlog::warn(
+					"ignoring Atlas connectionless packet (size={} type={}): invalid: invalid signature (key={})",
+					packet->size,
+					pType,
+					std::string(g_pMasterServerManager->m_sOwnServerAuthToken));
+			return;
+		}
+		spdlog::warn(
+			"processing Atlas connectionless packet (size={} type={}) with invalid signature due to net_debug_atlas_packet_insecure",
+			packet->size,
+			pType);
+	}
+
+	if (dbg)
+		spdlog::info("got Atlas connectionless packet (size={} type={} data={})", packet->size, pType, pData);
+
+	std::thread t(&MasterServerManager::ProcessConnectionlessPacketSigreq1, g_pMasterServerManager, pData);
+	t.detach();
+
+	return;
+}
+
+static void ProcessAtlasConnectionlessPacket(R2::netpacket_t* packet)
+{
+	bool dbg = Cvar_net_debug_atlas_packet->GetBool();
+
+	// extract kind, null-terminated type, data
+	std::string pType, pData;
+	for (int i = 5; i < packet->size; i++)
+	{
+		if (packet->data[i] == '\x00')
+		{
+			pType.assign((char*)(&packet->data[5]), (size_t)(i - 5));
+			if (i + 1 < packet->size)
+				pData.assign((char*)(&packet->data[i + 1]), (size_t)(packet->size - i - 1));
+			break;
+		}
+	}
+
+	// v1 HMACSHA256-signed masterserver request
+	if (pType == "sigreq1")
+	{
+		ProcessAtlasConnectionlessPacketSigreq1(packet, dbg, pType, pData);
+		return;
+	}
+
+	if (dbg)
+		spdlog::warn("ignoring Atlas connectionless packet (size={} type={}): unknown type", packet->size, pType);
+	return;
+}
+
+AUTOHOOK(ProcessConnectionlessPacket, engine.dll + 0x117800, bool, , (void* a1, R2::netpacket_t* packet))
+{
+	if (4 < packet->size && packet->data[4] == 'T')
+	{
+		ProcessAtlasConnectionlessPacket(packet);
+		return false;
+	}
+	if (!g_pServerLimits->CheckConnectionlessPacketLimits(packet))
+		return false;
+	return ProcessConnectionlessPacket(a1, packet);
+}
+
+ON_DLL_LOAD_RELIESON("engine.dll", ServerNetHooks, ConVar, (CModule module))
+{
+	AUTOHOOK_DISPATCH_MODULE(engine.dll)
+
+	if (!InitHMACSHA256())
+		throw std::runtime_error("failed to initialize bcrypt");
+
+	if (!VerifyHMACSHA256(
+			"test",
+			"\x88\xcd\x21\x08\xb5\x34\x7d\x97\x3c\xf3\x9c\xdf\x90\x53\xd7\xdd\x42\x70\x48\x76\xd8\xc9\xa9\xbd\x8e\x2d\x16\x82\x59\xd3\xdd"
+			"\xf7",
+			"test"))
+		throw std::runtime_error("bcrypt HMAC-SHA256 is broken");
+
+	Cvar_net_debug_atlas_packet = new ConVar(
+		"net_debug_atlas_packet",
+		"0",
+		FCVAR_NONE,
+		"Whether to log detailed debugging information for Atlas connectionless packets (warning: this allows unlimited amounts of "
+		"arbitrary data to be logged)");
+
+	Cvar_net_debug_atlas_packet_insecure = new ConVar(
+		"net_debug_atlas_packet_insecure",
+		"0",
+		FCVAR_NONE,
+		"Whether to disable signature verification for Atlas connectionless packets (DANGEROUS: this allows anyone to impersonate Atlas)");
+
+	if (Cvar_net_debug_atlas_packet_insecure->GetBool())
```

Comment:
Good catch; I haven't checked if that message actually shows yet. I'll probably just remove it since it'd get buried anyways, and we warn when receiving the packet too (plus nobody will ever know this exists unless they dig through the code).